### ISSUE TITLE
Asset ID

### ISSUE THREAD

**Issue by mottosso** [mindbender-studio/core#25]
### Goal

Enable directly referencing an `asset`, `subset`, `version` and `representation` via a short and unique identifier, such as `344574b2`.

<br>

### Motivation

`api.ls()` reads from disk and returns a fully populated dictionary for each asset, which leads to unnecessarily high disk usage.

<br>

### Implementation

This is the current interface.

```python
from mindbender import api

for asset in api.ls():
  print(asset["name"])
```

If you know the asset, there is currently no other way of receiving a dictionary for any particular asset, without looping through them.

```python
for asset in api.ls():
  if asset["name"] != "MyParticularAsset":
    continue

  assert asset["name"] == "MyParticularAsset"
```

This is how `search` currently works.

```python
# Version 3 of modelDefault for Bruce
asset = next(api.search("Bruce/modelDefault/3.ma"))
```

Worst case scenario, `ls` must loop through every asset on disk before finding the one you are looking for. Because each `asset` contains one or more `subset`, each subset contains one or more `versions` and each `version` one or more `representation`, this ultimately leads to a O(n^3) efficiency. In practice, a simple lookup on a regularly sized production set may take several seconds.

By instead keeping an index of every asset, subset, version and representation, then there would need to be no search and one could look up any of these components in O(1) time. To not depend on naming, an index could instead consist of a random string, such as (part of) a UUID.

```python
index = {
  "344574b2": {"schema": "asset", "name": "Bruce", "subsets": ["6434t54v", "fd353h1y"]}
  "7df16gdv": {"schema": "asset", "name": "Stella", "subsets": ["kr34j513", "3l25l464"]}
  "1hdf16gdv": {"schema": "version", "version": 4, "representations": ["u2hc513", "6u1l464"]}
}

# Instantly fetch a specific component
v004 = api.find("1hdf16gdv")
```

The index may then reside within a database, hosted either locally or even remotely (such as in Firebase). Locally is of course preferable for speed and reliability, but would require at least one dedicated computer with guaranteed up-time.

**Comment by mottosso** [comment]
Some progress on this.

**Proposed database layout**

```json
{
	"f034a4b": {
		"type": "asset",
		"name": "Bruce",
		"path": "{projectdir}/assets/{name}",
		"parent": null,
		"children": [
			"b07deyha",
			"u02f0dyh"
			"046kdaog"
		]
	},
	"b07deyha": {
		"type": "subset",
		"name": "modelDefault",
		"path": "{parentdir}/{name}",
		"parent": "l35dsog3",
		"children": [
			"od45qffl",
			"l35dsog3",
		]
	},
	"l35dsog3": {
		"type": "version",
		"name": "v001",
		"path": "{parentdir}/{name}",
		"parent": "b07deyha",
		"children": [
			"dre0463f",
			"ked4g9af",
		],
		"metadata": "{parentdir}/{name}/.metadata.json"
	},
	"dre0463f": {
		"type": "representation",
		"path": "{parentdir}/modelDefault.ma",
		"parent": "l35dsog3",
	}
}
```

This would require..

1. Upon creating a new project, a new database is created
2. Upon creating a new asset, an entry is made
3. For each published subset, an entry is made
4. For each version and representation, entries are made

Requirements 1 and 2 are currently not possible due to both projects and assets being created by primitive Windows-specific .bat files. What would need to happen is for the Terminal API to be written in a higher level language with the ability to open and modify the database which would be JSON for starters, Firebase later and eventually a self-hosted equivalent such as MongoDB.

**Comment by mottosso** [comment]
Implemented in #116.

Every project, asset, subset, version and representation now have a unique address accessible via `find`.

```python
from mindbender import io
asset = io.find({"_id": "unique_id_here}
asset["name"] == "Bruce"
```
